### PR TITLE
Issue #173 Rewrite NUMA CPU affinity parsing code for speed

### PR DESCRIPTION
**Pull request by pdbain-ibm**
Use the "cpumap" file if possible instead of parsing every entry in the
"node" directory.

Signed-off-by: Peter Bain <pdbain@ca.ibm.com>

### CODE REVIEW

File: thread/linux/omrthreadnuma.c [hunk 1]
```diff
@@ -134,42 +141,100 @@ initializeNumaNodeData(omrthread_library_t threadLibrary, uintptr_t numNodes)
...
+								size_t bytesRead = read(cpumapFile, mapBuffer, bufferSize);
+								if (bytesRead == bufferSize) { /* buffer possibly not big enough */
+									bufferSize = 4096;
+									mapBuffer = malloc(bufferSize);
+									if (NULL == mapBuffer) {
+										result = -1;
+										close(cpumapFile);
+										break;
+									}
+									size_t bytesRead = read(cpumapFile, mapBuffer, bufferSize);
```
Comment (DanHeidinga) [inline thread/linux/omrthreadnuma.c]: It seems pretty unlikely but what if bytesRead == bufferSize here?

File: thread/linux/omrthreadnuma.c [hunk 2]
```diff
@@ -134,42 +141,100 @@ initializeNumaNodeData(omrthread_library_t threadLibrary, uintptr_t numNodes)
...
+								}
+								if (0 != bytesRead) {
+									/* scan from the end of the line (lowest numbered CPU) toward the front */
+									char *cursor = mapBuffer+strlen(mapBuffer);
```
Comment (DanHeidinga) [inline thread/linux/omrthreadnuma.c]: Why strlen() rather than bytesRead?

File: thread/linux/omrthreadnuma.c [hunk 3]
```diff
@@ -23,16 +23,23 @@
...
+#include <sys/stat.h>
 #include <stdio.h>
 
 #include "omrcfg.h"
 #include "threaddef.h"
 
 static uintptr_t cpuset_subset_or_equal(cpu_set_t *subset, cpu_set_t *superset);
 static void cpuset_logical_or(cpu_set_t *destination, const cpu_set_t *source);
-
+#undef OMR_VERBOSE_INIT
```
Comment (DanHeidinga) [inline thread/linux/omrthreadnuma.c]: I might pick a more unique name for this.  Something like OMR_VERBOSE_NUMA_NODE_DATA?

File: thread/linux/omrthreadnuma.c (hunk 1 above)
Comment (pdbain-ibm) [inline thread/linux/omrthreadnuma.c]: I am in favour of throwing an assertion, but I could also fire  a tracepoint and return -1.
@charliegracie your thoughts?

File: thread/linux/omrthreadnuma.c (hunk 2 above)
Comment (pdbain-ibm) [inline thread/linux/omrthreadnuma.c]: That would assume that read() would return the string with no trailing crud, e.g. multiple nulls.
I should, however, read bufferSize-1 bytes and add a null at mapBuffer[bytesRead].

File: thread/linux/omrthreadnuma.c (hunk 3 above)
Comment (pdbain-ibm) [inline thread/linux/omrthreadnuma.c]: Sure. Will do.

**Comment by pdbain-ibm** [comment]
Changeset updated per Dan's comments, specifically:
> It seems pretty unlikely but what if bytesRead == bufferSize here?
in that case, fall back to scanning the directory.

**Pull request merged**